File: lib/mentions.ts
/**
 * `@mention` of an agent, as it is stored.
 *
 * ─────────────────────────────────────────────────────────────────────────
 *  THE WIRE FORMAT — this is the decision the backend has to live with.
 * ─────────────────────────────────────────────────────────────────────────
 *
 *   [@<slug>](crewship:agent/<agentId>)
 *
 * e.g.  `please pick this up [@robin](crewship:agent/cmt20ikph011ab4683c02)`
 *
 * It is a plain Markdown inline link. That one choice buys four properties
 * the alternatives do not:
 *
 *  1. **Unambiguously parseable.** The agent id lives in the link
 *     destination behind a private scheme nobody else emits, so a mention is
 *     `link.url.startsWith("crewship:agent/")` — one predicate, on a node the
 *     Markdown parser already produced. No hand-rolled scanner, and — because
 *     the parser produced it — a mention inside a code fence or a code span
 *     is *not* a link node and is therefore not a mention. That falls out for
 *     free; a regex over the raw body gets it wrong and turns documentation
 *     of the syntax into a live trigger.
 *
 *  2. **Degrades to readable text.** Nothing parses it? You read
 *     `[@robin](crewship:agent/cmt20…)` — the handle is right there, in
 *     front. Any generic Markdown renderer shows a link labelled `@robin`.
 *     Compare `<@cmt20ikph011ab4683c02>` (Slack's shape), which degrades to
 *     an opaque id and tells a human nothing.
 *
 *  3. **Addressed by id, never by name.** The only thing read out of a
 *     comment body is the agent id. The chip's name and avatar come from the
 *     workspace roster, resolved at render time. So no wording a user can
 *     type makes a chip *claim* anything: write
 *     `[@head-of-security](crewship:agent/<robin's id>)` and you get a chip
 *     reading "Robin", with Robin's face. This is the forgery defence, and it
 *     is why the label is decorative and the id is not.
 *
 *  4. **An id that does not resolve is not a mention.** It renders as plain
 *     muted text, never as a chip. So `[@ceo](crewship:agent/agt_made_up)`
 *     buys exactly as much as typing `@ceo` did: nothing.
 *
 * The honest limit: a body is text, so of course a human can *type* a
 * well-formed token for a real agent. That is not forgery — that is a
 * mention, which is the feature. What cannot be done is make a mention say
 * something the roster does not, or address an agent the reader's workspace
 * cannot see. Authorization ("may this actor make this agent work?") is the
 * backend's call at trigger time, and no wire format can stand in for it.
 *
 * ── What the backend has to implement to meet this ──────────────────────
 *
 *   - On comment write, parse the body with the equivalent of
 *     {@link extractMentionAgentIds} (the Go regexp is in the doc guide),
 *     resolve each id **within the comment's workspace**, and drop ids that
 *     do not resolve — a mention of a foreign-workspace agent is not a
 *     mention, it is a probe.
 *   - Persist the resolved set (a `mission_comment_mentions` join, or a
 *     column) rather than re-parsing on read: the body is user input and
 *     re-parsing on every read is a second chance to get it wrong.
 *   - Emit a `mentioned` activity per resolved mention, with `details` set to
 *     the agent id (the renderer also accepts a full token or
 *     `{"agent_id":"…"}` — see `mentionTargetFromActivityDetails`).
 *   - Only then dispatch the trigger, under the same authorization the
 *     equivalent "assign this agent" action would take.
 */

/** The private URL scheme a mention's destination must start with. */
export const MENTION_URL_SCHEME = "crewship:agent/"

/**
 * Ids we are willing to write into, or read out of, a comment body.
 *
 * CUIDs are `[a-z0-9]+`; this is deliberately a little wider (so a seeded or
 * prefixed id still round-trips) and deliberately not wide enough to hold a
 * slash, a dot, a space or a bracket. That is what stops a destination from
 * ever being a path (`../../etc/passwd`), a second token, or a URL to
 * somewhere else.
 */
const AGENT_ID_RE = /^[A-Za-z0-9_-]{1,64}$/

/** Characters a written label may contain. Everything else is dropped. */
const LABEL_ALLOWED_RE = /[^a-z0-9._-]+/g

/** Reading is liberal in the label, strict in the destination. */
const MENTION_TOKEN_RE = /\[@[^\]\n]{0,80}\]\(crewship:agent\/([A-Za-z0-9_-]{1,64})\)/g

/** Characters that may sit directly before the `@` that opens the picker. */
const HANDLE_CHAR_RE = /[a-zA-Z0-9._-]/

/**
 * An agent, as much of one as a mention needs. Structurally satisfied by
 * `AgentSummary` from the agents endpoint.
 */
export interface MentionAgent {
  id: string
  name: string
  slug: string
  /** DiceBear seed; falls back to the slug, then the name. */
  avatar_seed?: string | null
  avatar_style?: string | null
  avatar_url?: string | null
  role_title?: string | null
}

/** Agents by id — what the renderer resolves a mention against. */
export type MentionDirectory = ReadonlyMap<string, MentionAgent>

export function mentionDirectory(agents: readonly MentionAgent[]): MentionDirectory {
  return new Map(agents.map((a) => [a.id, a]))
}

/**
 * The label written into a token. Not identity — decoration. It is reduced
 * to the slug alphabet so it can never carry a `]` or a `)` and split one
 * token into two.
 */
export function normalizeMentionLabel(slug: string): string {
  const cleaned = slug.toLowerCase().replace(LABEL_ALLOWED_RE, "").slice(0, 64)
  return cleaned || "agent"
}

/** Write the token for an agent. Throws on an id this format cannot express. */
export function mentionToken(agent: Pick<MentionAgent, "id" | "slug" | "name">): string {
  if (!AGENT_ID_RE.test(agent.id)) {
    throw new Error(`mentionToken: agent id is not expressible in a mention: ${agent.id}`)
  }
  const label = normalizeMentionLabel(agent.slug || agent.name)
  return `[@${label}](${MENTION_URL_SCHEME}${agent.id})`
}

/**
 * The agent id behind a link destination, or null if this link is not a
 * mention. Case-sensitive on the scheme on purpose: `CREWSHIP:AGENT/` is not
 * a thing we write, so accepting it only widens what has to be reasoned about.
 */
export function parseMentionUrl(url: string | null | undefined): string | null {
  if (!url || !url.startsWith(MENTION_URL_SCHEME)) return null
  const id = url.slice(MENTION_URL_SCHEME.length)
  return AGENT_ID_RE.test(id) ? id : null
}

/** Every agent id mentioned in a body, de-duplicated, in first-seen order. */
export function extractMentionAgentIds(body: string): string[] {
  const seen = new Set<string>()
  const re = new RegExp(MENTION_TOKEN_RE.source, "g")
  let m: RegExpExecArray | null
  while ((m = re.exec(body)) !== null) seen.add(m[1])
  return [...seen]
}

/** An open `@…` at the caret: where it starts, and what has been typed. */
export interface MentionQuery {
  /** Index of the `@`. */
  start: number
  /** Lower-cased text between the `@` and the caret. */
  query: string
}

/**
 * Is the caret inside an `@handle` the picker should be open for?
 *
 * Only at a word boundary, so `pavel@unify.cz` is an email address and not a
 * half-typed mention of `unify.cz`.
 */
export function findMentionQuery(text: string, caret: number): MentionQuery | null {
  if (caret <= 0 || caret > text.length) return null
  let i = caret - 1
  while (i >= 0 && HANDLE_CHAR_RE.test(text[i])) i--
  if (i < 0 || text[i] !== "@") return null
  if (i > 0 && /\S/.test(text[i - 1])) return null
  return { start: i, query: text.slice(i + 1, caret).toLowerCase() }
}

/** Swap the typed `@handle` for the real token, and park the caret after it. */
export function applyMention(
  text: string,
  start: number,
  caret: number,
  agent: Pick<MentionAgent, "id" | "slug" | "name">,
): { text: string; caret: number } {
  const token = `${mentionToken(agent)} `
  const next = text.slice(0, start) + token + text.slice(caret)
  return { text: next, caret: start + token.length }
}

/**
 * The agent a `mentioned` activity points at.
 *
 * The activity kind does not exist server-side yet (see the guide), so this
 * accepts every shape the backend might plausibly put in `details` — a bare
 * id, a whole token, or a JSON object — rather than betting on one and
 * rendering nothing when the bet is wrong.
 */
export function mentionTargetFromActivityDetails(details: string | null): string | null {
  if (!details) return null
  const trimmed = details.trim()
  if (AGENT_ID_RE.test(trimmed)) return trimmed
  const [fromToken] = extractMentionAgentIds(trimmed)
  if (fromToken) return fromToken
  if (trimmed.startsWith("{")) {
    try {
      const parsed: unknown = JSON.parse(trimmed)
      if (parsed && typeof parsed === "object") {
        const id = (parsed as Record<string, unknown>).agent_id
        if (typeof id === "string" && AGENT_ID_RE.test(id)) return id
      }
    } catch {
      /* details is not JSON — nothing to read out of it. */
    }
  }
  return null
}

/** Activity `action` values the history row renders as "X mentioned @Y". */
const MENTION_ACTIONS = new Set(["mentioned", "mention", "agent_mentioned", "comment_mentioned"])

export function isMentionActivity(action: string): boolean {
  return MENTION_ACTIONS.has(action)
}

File: internal/mentions/mentions.go
// Package mentions reads agent @mentions out of a Markdown body.
//
// The wire format is a plain Markdown inline link behind a private scheme:
//
//	[@<label>](crewship:agent/<agentId>)
//
// The authoritative design rationale lives in `lib/mentions.ts` and the
// user-facing contract in `docs/guides/issue-mentions.mdx`. Two properties of
// that format are load-bearing here:
//
//  1. Only the id is identity. The label is decoration — the chip's name and
//     avatar come from the workspace roster at render time — so this package
//     returns ids and never labels. There is no API here that hands a caller a
//     display name read out of a comment body, because there must not be one.
//
//  2. A mention inside code is not a mention. Documenting the syntax, in an
//     issue comment, in a fenced block or a code span, must not fire an agent
//     run. This is a security property, not a cosmetic one: a trigger will
//     eventually hang off the result of ExtractAgentIDs.
//
// # Why a Markdown parser and not a regexp
//
// The doc guide currently suggests a bare regexp over the raw body:
//
//	regexp.MustCompile(`\[@[^\]\n]{0,80}\]\(crewship:agent/([A-Za-z0-9_-]{1,64})\)`)
//
// That regexp cannot see code, so it contradicts property 2 and the very table
// row in the same guide that promises "documenting this syntax does not fire a
// trigger". `extractMentionAgentIds` in `lib/mentions.ts` has the same hole; the
// frontend gets away with it because the thing that actually renders a chip
// (`components/features/issues/markdown-content.tsx`) walks the *parsed*
// document and only ever converts a link node.
//
// This package therefore does what the renderer does: parse the body with
// goldmark (CommonMark), walk the AST, and consider only `ast.Link` nodes whose
// destination is a well-formed mention URL. Code spans, fenced blocks, indented
// code blocks and raw HTML blocks contain no link nodes, so property 2 falls out
// of the parse instead of being a scanner rule somebody has to keep correct.
//
// goldmark is already in the module graph (glamour depends on it); no new
// third-party dependency was introduced.
//
// # What is guaranteed
//
//   - A returned id matched the strict id class `[A-Za-z0-9_-]{1,64}`, so it can
//     never be a path, a second token, or a URL to somewhere else. It is safe to
//     put in a SQL parameter, a URL path segment or a log line.
//   - Results are de-duplicated and in first-seen (document) order.
//   - Nothing lexically inside a fenced code block (``` or ~~~, any info string,
//     any fence length), an indented code block, an inline code span, or a raw
//     HTML block is ever returned.
//   - An unterminated code fence swallows the rest of the document, exactly as
//     CommonMark says it does. Mentions before the fence are still returned;
//     mentions after it are not. That is the fail-safe direction.
//
// # What is NOT guaranteed, and is deliberately out of scope
//
//   - Resolution. A returned id is a *claim*. It has not been checked against
//     any workspace. A mention of an agent in another workspace is a probe, not
//     a mention: the caller MUST resolve every id inside the comment's own
//     workspace and drop the ones that do not resolve.
//   - Authorization. Parsing says nothing about whether the comment's author may
//     make that agent work. Dispatch under the same authorization an "assign
//     this agent" action would take, never merely because the token parsed.
//   - Persistence and activity emission. Callers should persist the resolved set
//     rather than re-parsing on read.
//
// # Where this is deliberately stricter than the frontend
//
// Where the doc guide's regexp and the frontend renderer disagree, this package
// takes the intersection — it fires only where *both* would. The failure mode is
// "miss a mention", never "fire on documentation":
//
//   - The label must conform to the doc guide's regexp: it must start with `@`,
//     and the remainder must be at most 80 bytes with no `]`, `\n` or `\r`. The
//     renderer would happily chip a link whose text spans a line or contains a
//     bracket; this package will not.
//   - A link whose label cannot be located in the source (an autolink or an
//     HTML entity inside the label, neither of which carries a source position)
//     is rejected rather than guessed at.
//   - An empty label (`[](crewship:agent/…)`) is rejected. The renderer chips it,
//     the doc guide's regexp does not, and a mention nobody can read is not one.
//   - An image (`![@robin](crewship:agent/…)`) is not a mention.
//
// # Where it is deliberately wider than the doc guide's regexp
//
// Both of these are real link nodes, so the reader genuinely sees a chip; the
// regexp missing them is the regexp's bug, not a safety margin:
//
//   - A reference-style link (`[@robin][r]` plus `[r]: crewship:agent/…`).
//     A bare definition with no reference to it produces no link node and so is
//     not a mention.
//   - Whitespace inside the destination parens (`[@robin]( crewship:agent/… )`).
package mentions

import (
	"strings"

	"github.com/yuin/goldmark"
	"github.com/yuin/goldmark/ast"
	"github.com/yuin/goldmark/text"
)

// URLScheme is the private URL scheme a mention's destination must start with.
// Case-sensitive on purpose: `CREWSHIP:AGENT/` is not a thing we ever write, so
// accepting it only widens what has to be reasoned about.
const URLScheme = "crewship:agent/"

// MaxIDLen is the longest agent id a mention can express.
const MaxIDLen = 64

// MaxLabelLen is the longest label (excluding the leading `@`) a mention can
// carry. Taken from the doc guide's regexp, not invented here.
const MaxLabelLen = 80

// mdParser is CommonMark with no extensions. goldmark guards its one-time
// initialisation with a sync.Once and allocates a fresh parse context per call,
// so a package-level parser is safe for concurrent use.
var mdParser = goldmark.New().Parser()

// ValidID reports whether id is expressible as a mention destination.
//
// The class is deliberately narrow — CUIDs are `[a-z0-9]+` and this is a little
// wider so a seeded or prefixed id still round-trips, but not wide enough to
// hold a slash, a dot, a space or a bracket. That is what stops a destination
// from ever being a path (`../../etc/passwd`), a second token, or a URL
// somewhere else.
func ValidID(id string) bool {
	if len(id) == 0 || len(id) > MaxIDLen {
		return false
	}
	for i := 0; i < len(id); i++ {
		c := id[i]
		switch {
		case c >= 'a' && c <= 'z',
			c >= 'A' && c <= 'Z',
			c >= '0' && c <= '9',
			c == '_', c == '-':
		default:
			return false
		}
	}
	return true
}

// ParseURL returns the agent id behind a link destination, and whether the
// destination is a mention at all. It does not resolve the id.
func ParseURL(dest string) (string, bool) {
	if !strings.HasPrefix(dest, URLScheme) {
		return "", false
	}
	id := dest[len(URLScheme):]
	if !ValidID(id) {
		return "", false
	}
	return id, true
}

// ExtractAgentIDs returns every agent id mentioned in a Markdown body,
// de-duplicated and in first-seen order.
//
// The ids are unresolved claims: see the package doc for what the caller still
// owes before any of them may trigger work. Anything inside code is ignored.
func ExtractAgentIDs(body string) []string {
	if body == "" {
		return nil
	}

	src := []byte(body)
	doc := mdParser.Parse(text.NewReader(src))

	var ids []string
	seen := make(map[string]struct{})

	// ast.Walk only returns an error if the callback does; this one never does.
	_ = ast.Walk(doc, func(n ast.Node, entering bool) (ast.WalkStatus, error) {
		if !entering {
			return ast.WalkContinue, nil
		}
		link, ok := n.(*ast.Link)
		if !ok {
			return ast.WalkContinue, nil
		}
		id, ok := ParseURL(string(link.Destination))
		if !ok {
			return ast.WalkContinue, nil
		}
		if !labelConforms(link, src) {
			return ast.WalkContinue, nil
		}
		if _, dup := seen[id]; dup {
			return ast.WalkContinue, nil
		}
		seen[id] = struct{}{}
		ids = append(ids, id)
		return ast.WalkContinue, nil
	})

	return ids
}

// labelConforms reports whether a link's text, as it was written in the source,
// satisfies the `\[@[^\]\n]{0,80}\]` half of the specified token shape.
//
// The check runs on the raw source span the label occupies rather than on the
// decoded text, so it sees what the doc guide's regexp would have seen. If the
// span cannot be determined the link is rejected — see the package doc.
func labelConforms(link *ast.Link, src []byte) bool {
	start, stop, ok := sourceSpan(link)
	if !ok || start < 0 || stop > len(src) || start >= stop {
		return false
	}
	label := src[start:stop]
	if label[0] != '@' {
		return false
	}
	rest := label[1:]
	if len(rest) > MaxLabelLen {
		return false
	}
	for _, c := range rest {
		if c == ']' || c == '\n' || c == '\r' {
			return false
		}
	}
	return true
}

// sourceSpan returns the half-open source range a node's children occupy.
//
// ok is false when any descendant carries no source position (an autolink, or an
// HTML entity, both of which goldmark materialises without a segment). Callers
// treat that as "cannot verify" and reject, rather than measuring a span that is
// narrower than what was actually written.
func sourceSpan(n ast.Node) (start, stop int, ok bool) {
	child := n.FirstChild()
	if child == nil {
		return 0, 0, false
	}
	start, stop = -1, -1
	for ; child != nil; child = child.NextSibling() {
		s, e, childOK := nodeSpan(child)
		if !childOK {
			return 0, 0, false
		}
		if start < 0 || s < start {
			start = s
		}
		if e > stop {
			stop = e
		}
	}
	return start, stop, start >= 0 && stop > start
}

// nodeSpan returns the source range a single inline node occupies.
func nodeSpan(n ast.Node) (start, stop int, ok bool) {
	switch v := n.(type) {
	case *ast.Text:
		return v.Segment.Start, v.Segment.Stop, true
	case *ast.RawHTML:
		if v.Segments == nil || v.Segments.Len() == 0 {
			return 0, 0, false
		}
		first := v.Segments.At(0)
		last := v.Segments.At(v.Segments.Len() - 1)
		return first.Start, last.Stop, true
	default:
		// Emphasis, code spans, nested links: recurse into the children, which
		// bottom out at Text nodes.
		return sourceSpan(n)
	}
}
